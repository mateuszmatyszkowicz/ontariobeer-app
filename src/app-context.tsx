import React, { createContext, useEffect, useReducer } from "react";

export type Brewer = {
  productId: number;
  name: string;
  size: string;
  price: number;
  thumbnail: string;
  type: string;
  brewer: string;
};

export type AppState = {
  themeMode: "light" | "dark";
  sortBy: "type" | "name" | "price";
  itemsPerPage: number;
  brewers?: string[];
  byBrewers?: Record<string, Brewer[]>;
  loading: boolean;
  error?: string;
};

export const initialState: AppState = {
  themeMode: "light",
  sortBy: "type",
  itemsPerPage: 15,
  brewers: [],
  byBrewers: {},
  loading: false,
  error: undefined,
};

export const stateInitializer = (initialValue = initialState): AppState => {
  const themeMode =
    (localStorage.getItem("themeMode") as AppState["themeMode"]) ||
    initialValue.themeMode;

  return {
    ...initialValue,
    themeMode,
  };
};

export const appReducer = (state: AppState, action: any): AppState => {
  switch (action.type) {
    case "app/SET_ITEMS_PER_PAGE": {
      return {
        ...state,
        itemsPerPage: action.payload.itemsPerPage,
      };
    }
    case "app/SET_THEME":
      return {
        ...state,
        themeMode: action.payload.theme,
      };
    case "app/SET_LOADING":
      return {
        ...state,
        loading: true,
      };
    case "app/SET_ERROR":
      return {
        ...state,
        loading: false,
        error: "Something went wrong...",
      };
    case "app/SET_DATA":
      const byBrewers = action.payload.data
        ?.map(
          (data: any) =>
            ({
              type: data.type,
              brewer: data.brewer,
              productId: data.product_id,
              name: data.name,
              price: data.price,
              thumbnail: data.image_url,
              size: data.size,
            } as Brewer)
        )
        ?.reduce(
          //@ts-ignore
          (prev, curr) => {
            if (curr.brewer && prev[curr.brewer]) {
              return {
                ...prev,
                [curr.brewer]: [...prev[curr.brewer], curr],
              };
            } else {
              return {
                ...prev,
                [curr.brewer]: [curr],
              };
            }
          },
          {}
        );

      return {
        ...state,
        loading: false,
        byBrewers,
        brewers: Object.keys(byBrewers),
      };
    default:
      throw new Error();
  }
};

//@ts-ignore
export const AppContext = createContext<[AppState, React.Dispatch<any>]>({});
export const AppContextProvider: React.FC = ({ children }) => {
  const [state, dispatch] = useReducer(
    appReducer,
    initialState,
    stateInitializer
  );
  const { themeMode, itemsPerPage } = state;
  // Perhaps we could serialize whole appState and save it to localStorage some window event.
  useEffect(() => {
    localStorage.setItem("themeMode", themeMode);
  }, [themeMode]);

  useEffect(() => {
    localStorage.setItem("itemsPerPage", itemsPerPage.toString());
  }, [itemsPerPage]);

  return (
    <AppContext.Provider value={[state, dispatch]}>
      {children}
    </AppContext.Provider>
  );
};

export const setLoading = () => ({ type: "app/SET_LOADING" });
export const setError = () => ({ type: "app/SET_ERROR" });
export const setData = (data: any) => ({
  type: "app/SET_DATA",
  payload: { data },
});
export const setTheme = (theme: "light" | "dark") => ({
  type: "app/SET_THEME",
  payload: { theme },
});
export const setItemsPerPage = (itemsPerPage: number) => ({
  type: "app/SET_ITEMS_PER_PAGE",
  payload: { itemsPerPage },
});
