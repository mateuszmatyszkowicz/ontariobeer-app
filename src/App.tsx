import React, { useContext, useEffect } from "react";
import { AppContext, setLoading, setError, setData } from "./app-context";
import Column from "./column.component";
import "./App.css";
import Settings from "./settings.component";

function App() {
  const [appState, dispatch] = useContext(AppContext);

  useEffect(() => {
    const fetchBeers = async () => {
      try {
        dispatch(setLoading());
        const data = await fetch("/beers");
        const jsonData = await data.json();

        dispatch(setData(jsonData));
      } catch (e) {
        dispatch(setError());
      }
    };

    fetchBeers();
    //eslint-disable-next-line
  }, []);

  return (
    <div className="App">
      <Settings />
      <h1>Frontend developer coding test</h1>
      {appState.loading && <div>loading</div>}
      {appState.error && <div>{appState.error}</div>}
      {!!appState?.brewers?.length && (
        <div>
          <Column id={"1"} />
          <Column id={"2"} />
          <Column id={"3"} />
        </div>
      )}
    </div>
  );
}

export default App;
