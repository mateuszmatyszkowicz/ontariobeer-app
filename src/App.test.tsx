import React from "react";
import {
  cleanup,
  render,
  waitFor,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import App from "./App";

// eslint-disable-next-line
import beersApiResponse from "./__mocks__/beers-api-response.json";
import { AppContextProvider } from "./app-context";

const BREWER_OPTIONS = [
  "Molson",
  "Lake Of Bays Brewing",
  "Hockley Valley Brewing Company",
];

beforeEach(() => {
  cleanup();
  // @ts-ignore
  jest.spyOn(global, "fetch").mockImplementation(() =>
    Promise.resolve({
      json: () => Promise.resolve(beersApiResponse),
    })
  );
});


function renderApp() {
  return render(
    <AppContextProvider>
      <App />
    </AppContextProvider>
  );
}

test("renders application", async () => {
  const { findByText } = renderApp();

  const headerContent = await findByText(/frontend developer coding test/i);
  expect(headerContent).toBeInTheDocument();
});

test("should call api at first render", async () => {
  renderApp();

  await waitFor(() => expect(fetch).toBeCalledWith("/beers"));
  await waitFor(() => expect(fetch).toBeCalledTimes(1));
});

test("should render loading message", async () => {
  const { findByText } = renderApp();
  const loading = await findByText(/loading/i);

  expect(loading).toBeDefined();
});

test("should reder error message", async () => {
  // @ts-ignore
  jest.spyOn(global, "fetch").mockImplementation(() =>
    Promise.reject({
      json: () => Promise.reject(beersApiResponse),
    })
  );
  const { findByText, queryByText } = renderApp();
  await waitForElementToBeRemoved(() => queryByText("loading"));
  await waitFor(() => expect(fetch).toBeCalledWith("/beers"));
  await findByText("Something went wrong...");
});

test("should render options based on fetch data", async () => {
  const { findByTestId, findByText } = renderApp();

  const visibleLoading = await findByText("loading");
  const visibleColumn = await findByTestId("column-1");
  const visibleOptions = await visibleColumn.getElementsByTagName("option");
  expect(visibleOptions[0].value).toEqual(BREWER_OPTIONS[0]);
  expect(visibleOptions[1].value).toEqual(BREWER_OPTIONS[1]);
  expect(visibleOptions[2].value).toEqual(BREWER_OPTIONS[2]);
  expect(visibleLoading).not.toBeInTheDocument();
});
