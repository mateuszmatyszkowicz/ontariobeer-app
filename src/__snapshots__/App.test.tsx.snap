// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders application 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <h1>
      Frontend developer coding test
    </h1>
    <div>
      <div
        data-testid="column-1"
      >
        <select
          data-testid="select"
        >
          <option
            disabled=""
            hidden=""
            selected=""
            value=""
          >
            Select your brewery
          </option>
        </select>
        <div
          class="column-content"
        />
      </div>
      <div
        data-testid="column-2"
      >
        <select
          data-testid="select"
        >
          <option
            disabled=""
            hidden=""
            selected=""
            value=""
          >
            Select your brewery
          </option>
        </select>
        <div
          class="column-content"
        />
      </div>
      <div
        data-testid="column-3"
      >
        <select
          data-testid="select"
        >
          <option
            disabled=""
            hidden=""
            selected=""
            value=""
          >
            Select your brewery
          </option>
        </select>
        <div
          class="column-content"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
